spec/models/account_spec.rb

```diff
@@ -362,42 +362,99 @@
+      describe 'username should be present' do
+        let(:username) { nil }
+        it { is_expected.not_to be_valid }
+      end
+      describe 'username should be unique' do
+        before { Fabricate(:account, username: username, domain: domain) }
+        it { is_expected.not_to be_valid }
+      end
+    end
+
+    shared_examples 'not allowed duplicate username' do
+      before { Fabricate(:account, username: 'TootSuite', domain: domain) }
+      it { is_expected.to be_valid }
+
+      context 'when the username is already exists' do
+        let(:username) { 'TootSuite' }
+        it { is_expected.not_to be_valid }
+      end
+    end
+
+    shared_examples 'not enabled for existing record' do
+      before { subject.save!(validate: false) }
+      it { is_expected.to be_valid }
+    end
+
+    it_behaves_like 'check username validation'
+    it_behaves_like 'not allowed duplicate username'
+
+    context 'when username contains invalid character' do
+      let(:username) { 'the-doctor' }
+      it { is_expected.not_to be_valid }
+      it_behaves_like 'not enabled for existing record'
     end
 
-    it 'is invalid without a username' do
-      account = Fabricate.build(:account, username: nil)
-      account.valid?
-      expect(account).to model_have_error_on_field(:username)
+    context 'when username is longer than 30 characters' do
+      let(:username) { Faker::Lorem.characters(31) }
+      it { is_expected.not_to be_valid }
+      it_behaves_like 'not enabled for existing record'
     end
 
-    it 'is invalid if the username already exists' do
-      account_1 = Fabricate(:account, username: 'the_doctor')
-      account_2 = Fabricate.build(:account, username: 'the_doctor')
-      account_2.valid?
-      expect(account_2).to model_have_error_on_field(:username)
+    context 'when display name is longer than 30 characters' do
+      let(:display_name) { Faker::Lorem.characters(31) }
+      it { is_expected.not_to be_valid }
+      it_behaves_like 'not enabled for existing record'
     end
 
-    it 'is invalid if the username is reserved' do
-      account = Fabricate.build(:account, username: 'support')
-      account.valid?
-      expect(account).to model_have_error_on_field(:username)
+    context 'when note is longer than 160 characters' do
+      let(:note) { Faker::Lorem.characters(161) }
+      it { is_expected.not_to be_valid }
+      it_behaves_like 'not enabled for existing record'
     end
 
-    context 'when is local' do
-      it 'is invalid if the username doesn\'t only contains letters, numbers and underscores' do
-        account = Fabricate.build(:account, username: 'the-doctor')
-        account.valid?
-        expect(account).to model_have_error_on_field(:username)
+    describe 'reserved username' do
+      let(:username) { 'support' }
+
+      it { is_expected.not_to be_valid }
+      it_behaves_like 'not enabled for existing record'
+    end
+
+    context 'local? => false' do
```

Comment:
I would just call this `context 'A remote account'` or something, instead of the `local?` impl detail.